Allow terminal flow state handler override (#4311)
PrefectHQ/prefect: docs/core/concepts/flows.md

The handler is called anytime the flow's state changes, and receives the flow itself, the old state, and the new state. The state that the handler returns is used as the flow's new state.

If multiple handlers are provided, they are called in sequence. Each one will receive the "true" `old_state` and the `new_state` generated by the previous handler.

## Terminal State Handlers

> Flows allow users to provide custom logic for determining the final State of a Flow.

Reviewer: ```suggestion
Flows allow users to provide custom logic for determining the final State of a Flow.
Unlike normal handlers, the terminal handler has access to the final states of all of the tasks in the flow.
```

Terminal state handlers must have the following signature:

```python
terminal_state_handler(flow: Flow, state: State, task_states: Dict[Task, State]) -> Optional[State]
```

`flow` is the current Flow
`state` is the current state of the Flow
`task_states` contains states for Flow Tasks

> An example use case would be iterating through return states and indicating which reference tasks that have failed.

Reviewer: ```suggestion
For example, you may want to determine which reference tasks that have failed and add them to your final state message.
```

Reviewer: Note that this is particularly useful for Cloud Hooks (/ Automations) which report the state message. Now users can customize the success/failure message so their notifications include useful information about what has failed.

The handler is called anytime the flow's state changes, and receives the flow itself, the old state, and the new state. The state that the handler returns is used as the flow's new state.

If multiple handlers are provided, they are called in sequence. Each one will receive the "true" `old_state` and the `new_state` generated by the previous handler.

## Terminal State Handlers

> Flows allow users to provide custom logic for determining the final State of a Flow.
> 
> Unlike normal state handlers, the terminal handler has access to the final states of all tasks in the flow.

Reviewer 2: ```suggestion
A flow's `terminal_state_handler` allow users to provide custom logic for determining the
final State of a Flow. Unlike normal state handlers, the `terminal_state_handler` has access
to the final states of all tasks in the flow.
```

Terminal state handlers must have the following signature:

```python
terminal_state_handler(flow: Flow, state: State, task_states: Dict[Task, State]) -> Optional[State]
```

> 
> `flow` is the current Flow
> `state` is the current state of the Flow
> `task_states` contains states for Flow Tasks

Reviewer 2: Single newlines are ignored in markdown, you probably want this to be a list:

```suggestion
where:

- `flow` is the current Flow
- `state` is the current state of the Flow
- `task_states` contains states for Flow Tasks
```

> For example, you may want to determine which reference tasks that have failed and add them to your final state message.
> 
> This is particularly useful for Cloud Hooks that report the state message.

Reviewer 2: ```suggestion
For example, you may want to determine which reference tasks have failed and add them
to your final state message. This could be used to improve the message sent out by Cloud
Hooks that report the flow run's final state.
```

```python
def custom_terminal_state_handler(
    flow: Flow,
    state: State,
    task_states: Dict[Task, State],
) -> Optional[State]:
    # iterate through task states, making a list of failing refernce tasks
    failed_tasks = []
    for task, task_state in task_states.items():
        if task_state.is_failed() and task in flow.reference_tasks():
            failed_tasks.append(task.name)
    # update the terminal state of the Flow and return
    state.message = "The following tasks failed: {}".format(failed_tasks)
    return state

# create a new flow using the terminal state handler

> f = Flow("my flow with custom terminal state handler", terminal_state_handler=custom_terminal_state_handler)

Reviewer 2: ```suggestion
flow = Flow(
    "my flow with custom terminal state handler",
    terminal_state_handler=custom_terminal_state_handler
)
```
